CRI: add ImageFsInfo API

feiskyer opened the pull request:
**What this PR does / why we need it**:

kubelet currently relies on cadvisor to get the ImageFS info for supported runtimes, i.e., docker and rkt. This PR adds ImageFsInfo API to CRI so kubelet could get the ImageFS correctly for all runtimes.

**Which issue this PR fixes** 

First step for  #33048 ~~also reverts temporary ImageStats in #33870~~.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41733)
<!-- Reviewable:end -->

euank commented:
the e2e failures do look related

Random-Liu commented:
@feiskyer I remember that at that time we decided to punt this change because the container metrics is not defined yet. We may need resource monitoring team's opinion for this. :) @kubernetes/sig-node-pr-reviews

feiskyer commented:
Relate with #33048. Need opinions from @kubernetes/sig-node-pr-reviews

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    // Path where the filesystem is mounted.
+    string mountpoint = 2;
+    // Filesystem usage in bytes.
+    uint64 capacity = 3;
```
rename it to `capacity_bytes`

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    string device = 1;
+    // Path where the filesystem is mounted.
+    string mountpoint = 2;
+    // Filesystem usage in bytes.
```
Comment is wrong; should be capacity.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    string mountpoint = 2;
+    // Filesystem usage in bytes.
+    uint64 capacity = 3;
+    // Bytes available for non-root use.
```
Is `non-root` meaningful here? I'd rather remove that from the comment.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    // Filesystem usage in bytes.
+    uint64 capacity = 3;
+    // Bytes available for non-root use.
+    uint64 available = 4;
```
rename this to `available_bytes`

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    // Bytes available for non-root use.
+    uint64 available = 4;
+    // Number of bytes used on this filesystem.
+    uint64 usage = 5;
```
rename this to `usage_bytes`

Random-Liu replied:
Why for disk usage, it's `capacity`, `available` and `usage`; but for inode, it's `used`, `available` and `free`?

We should make them consistent.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    uint64 available = 4;
+    // Number of bytes used on this filesystem.
+    uint64 usage = 5;
+    // Labels associated with this filesystem.
```
I don't think we need to expose labels, which is a very cadvisor-specific concept.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    // Labels associated with this filesystem.
+    repeated string labels = 6;
+    // Number of Inodes.
+    UInt64Value inodes = 7;
```
Why using `UInt64Value`? Is this optional? If inode information *is* optional, I'd suggest grouping them into a separate message.

Also rename the fields to:
`inodes_used`
`inodes_available`
`inodes_free`

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -90,6 +90,8 @@ service ImageService {
     // This call is idempotent, and must not return an error if the image has
     // already been removed.
     rpc RemoveImage(RemoveImageRequest) returns (RemoveImageResponse) {}
+    // ImageFsInfo returns filesytem information of images.
```
typo: s/filesytem/filesystem

// ImageFSInfo returns information of the filesystem that is used to store images.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+    // The block device name associated with the filesystem.
+    string device = 1;
+    // Path where the filesystem is mounted.
+    string mountpoint = 2;
```
Maybe just the `path` is enough? Do we really need to know the mount point, or just root directory for the images? I think latter is sufficient. 

/cc @vishh

dashpole replied:
Can you clarify what you mean by root directory for the images?  How is this different from mountpoint? An example would help.  :)

feiskyer replied:
I think @yujuhong also meant the mount point here, but with a more general naming.

yujuhong reviewed (commented):
I think we should nail down the API first, before plumbing it through in kubelet. @feiskyer could you leave out the implementation part in this PR? I want reviewers to focus on the API first.

@dashpole @vishh @dchen1107 @Random-Liu PTAL.

feiskyer commented:
Addressed comments and only keeping CRI API. Some highlights:

- Since all fields in [FsStats](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go#L201) are optional, new commits also made them optional.
- Make field names consistent for bytes and inodes.
- Update comments according to [FsStats](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go#L201).

@yujuhong @Random-Liu @vishh @dashpole @dchen1107  PTAL

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+    // The block device name associated with the filesystem.
+    string device = 1;
+    // The root directory for the images.
+    string path = 2;
```
I would prefer a more descriptive name here.  Path is a little generic.  I believe we generally refer to this as mountpoint in the rest of the code base.

dashpole replied:
Ignore this for now, just saw yuju's comment.

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+    // InodesAvailable represents the free inodes in the filesystem.
+    UInt64Value inodes_available = 7;
+    // InodesUsed represents the inodes used by the filesystem
+    // This may not equal InodesCapacity - InodesAvailable because this filesystem
```
Correct me if I am wrong, but I believe inodes are a filesystem construct, meaning that they cant be "shared".   That should mean that Capacity - Available ~= Used.

feiskyer replied:
@dashpole Yes, of course. I think your statements are exactly same with the comments here.

dashpole replied:
Sorry, I meant Capacity - Available approximately equals Used.  So the opposite of the comment.  AFAIK Inodes cannot be shared between filesystems.

feiskyer replied:
It could be shared between host and containers, or multiple containers.

Refer https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume.go#L75

dashpole replied:
Right, those metrics are for a **user** of a filesystem.  Users of a filesystem share the filesystem with other users.
Here, we are describing the filesystem itself.  I took usage to mean total usage of the filesystem.  "inodes used by the filesystem" should be "inodes consumed in the filesystem".

Random-Liu replied:
It seems that `InodesUsed` here should be "inode used by image".

Image file system is just a partition images lie in, the file system may be used by other programs, too.

`InodesAvailable` should be `(InodesCapacity-InodesUsed-InodesUsedByOtherPrograms)`.

dashpole replied:
Ok, that makes sense then.  I didn't realize "usage" referred to image usage.  Thanks for clarifying @Random-Liu.
When we [compute the summary metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/summary.go#L152), we populate AvailableBytes, CapacityBytes, InodesFree, Inodes, and InodesUsed using the ImageFsInfo provided by cAdvisor.  We populate UsedBytes using runtime.ImageStats(), not using ImageFsInfo.  Hence my confusion.
I agree with @Random-Liu that we should change the description to `inodes used by images`.

feiskyer replied:
@dashpole @Random-Liu Thanks for the review. Updated.

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+    // filesystem.
+    UInt64Value available_bytes = 4;
+    // UsedBytes represents the bytes used for images on the filesystem.
+    // This may differ from the total bytes used on the filesystem and may not 
```
same as below.

feiskyer commented:
@Random-Liu @dashpole @yujuhong Updated. PTAL.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,36 @@ message StatusResponse {
+
+message ImageFsInfoResponse {
+    // filesystem information of images.
+    FsInfo fs_info = 1;
```
Ping @yujuhong @vishh @dchen1107 Do you think we need to add multiple image filesystems now?

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+    UInt64Value inodes_available = 7;
+    // InodesUsed represents the inodes used by the images.
+    // This may not equal InodesCapacity - InodesAvailable because this filesystem
+    // may share inodes with other "filesystems"
```
I would prefer something more like:
...`because the underlying filesystem may also be used for purposes other than storing images`

feiskyer replied:
ack

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+    // This may differ from the total bytes used on the filesystem and may not 
+    // equal CapacityBytes - AvailableBytes.
+    UInt64Value used_bytes = 5;
+    // InodesCapacity represents the total inodes in the filesystem.
```
Is there any reason why this is camelcase, and the actual name uses underscores?

feiskyer replied:
It is for generating golang consistent comments.

dashpole commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -970,3 +972,50 @@ message StatusResponse {
+}
+
+// InodeInfo is the inode information of the images.
+message InodeInfo {
```
I cant find where this is used.  What is this for?

feiskyer replied:
Oops, forgotten to remove this.

feiskyer commented:
@dashpole Addressed comments. PTAL.

feiskyer commented:
ping @yujuhong @Random-Liu @vishh @dashpole @dchen1107

yujuhong commented:
Looks good to me overall. @vishh @dchen1107 do you want to take a look?

feiskyer committed:
CRI: Add ImageFsInfo API

feiskyer committed:
Run hack/update-generated-runtime.sh

feiskyer committed:
Add ImageFsInfo API for ImageManagerService

feiskyer commented:
Made a rebase.

feiskyer commented:
@k8s-bot unit test this

feiskyer commented:
Another rebase.

feiskyer commented:
@k8s-bot unit test this

dashpole commented:
Other than whether or not we plan to handle multiple filesystems, this looks good to me!

yujuhong commented:
/lgtm
